File: velocity/src/main/java/dev/rafo/bedrockbridge/velocity/hud/RestoringHudSession.java
package dev.rafo.bedrockbridge.velocity.hud;

import java.util.EnumSet;
import java.util.Set;
import java.util.concurrent.atomic.AtomicBoolean;

public final class RestoringHudSession implements AutoCloseable {
    private static final Set<HudElement> CINEMATIC_ELEMENTS = Set.copyOf(EnumSet.allOf(HudElement.class));

    private final HudController controller;
    private final Set<HudElement> originallyHidden;
    private final AtomicBoolean closed = new AtomicBoolean();

    public RestoringHudSession(HudController controller) {
        this.controller = controller;
        EnumSet<HudElement> original = copyOf(controller.hiddenElements());
        original.retainAll(CINEMATIC_ELEMENTS);
        originallyHidden = Set.copyOf(original);
        keepActive();
    }

    public void keepActive() {
        if (closed.get()) {
            return;
        }
        EnumSet<HudElement> missing = EnumSet.copyOf(CINEMATIC_ELEMENTS);
        missing.removeAll(controller.hiddenElements());
        if (!missing.isEmpty()) {
            controller.hide(missing);
        }
    }

    @Override
    public void close() {
        if (!closed.compareAndSet(false, true)) {
            return;
        }

        EnumSet<HudElement> addedByBridge = EnumSet.copyOf(CINEMATIC_ELEMENTS);
        addedByBridge.removeAll(originallyHidden);
        try {
            if (!addedByBridge.isEmpty()) {
                controller.reset(addedByBridge);
            }
        } finally {
            EnumSet<HudElement> hiddenBefore = copyOf(originallyHidden);
            hiddenBefore.removeAll(controller.hiddenElements());
            if (!hiddenBefore.isEmpty()) {
                controller.hide(hiddenBefore);
            }
        }
    }

    private static EnumSet<HudElement> copyOf(Set<HudElement> elements) {
        return elements.isEmpty() ? EnumSet.noneOf(HudElement.class) : EnumSet.copyOf(elements);
    }
}

File: src/main/kotlin/dev/rafo/bedrockbridge/geyser/RestoringHudSession.kt
package dev.rafo.bedrockbridge.geyser

import dev.rafo.bedrockbridge.state.CinematicSession
import java.util.concurrent.atomic.AtomicBoolean

internal enum class HudElement {
    PAPER_DOLL,
    ARMOR,
    TOOL_TIPS,
    TOUCH_CONTROLS,
    CROSSHAIR,
    HOTBAR,
    HEALTH,
    PROGRESS_BAR,
    FOOD_BAR,
    AIR_BUBBLES_BAR,
    VEHICLE_HEALTH,
    EFFECTS_BAR,
    ITEM_TEXT_POPUP,
}

internal interface HudCamera {
    fun hiddenElements(): Set<HudElement>

    fun hide(elements: Set<HudElement>)

    fun reset(elements: Set<HudElement>)
}

internal class RestoringHudSession(
    private val camera: HudCamera,
) : CinematicSession {
    private val closed = AtomicBoolean(false)
    private val originallyHidden = camera.hiddenElements().intersect(CINEMATIC_HUD_ELEMENTS)

    init {
        keepActive()
    }

    override fun keepActive() {
        if (closed.get()) return
        val missing = CINEMATIC_HUD_ELEMENTS - camera.hiddenElements()
        if (missing.isNotEmpty()) camera.hide(missing)
    }

    override fun close() {
        if (!closed.compareAndSet(false, true)) return

        val addedByBridge = CINEMATIC_HUD_ELEMENTS - originallyHidden
        try {
            if (addedByBridge.isNotEmpty()) camera.reset(addedByBridge)
        } finally {
            val hiddenBeforeCinematic = originallyHidden - camera.hiddenElements()
            if (hiddenBeforeCinematic.isNotEmpty()) camera.hide(hiddenBeforeCinematic)
        }
    }

    private companion object {
        val CINEMATIC_HUD_ELEMENTS = HudElement.entries.toSet()
    }
}
